onnxruntime/core/providers/cuda/tensor/transpose.h

```diff
@@ -12,12 +12,14 @@
 namespace onnxruntime {
 namespace cuda {
 
-template <typename T>
 class Transpose final : public CudaKernel, public TransposeBase {
  public:
   Transpose(const OpKernelInfo& info) : CudaKernel(info), TransposeBase(info) {}
 
   Status ComputeInternal(OpKernelContext* context) const override;
+
+  static Status DoTranspose(const Transpose& transpose_kernel,
+                            const std::vector<size_t>& permutations, const Tensor& input, Tensor& output);
 };
```

Comment:
The CUDA friendly Scan (current WIP) needs to be able to re-use the transpose logic so separate it out so it can be called directly. 

The Scan implementation will construct a Transpose kernel as it needs internals from that, and pass it in. It will not go via ComputeInternal as the OpKernelContext it has will be for a Scan op not a Transpose op.